File: src/go/problems/IsSubsequence_test.go
//https://leetcode.com/problems/is-subsequence/
package problems

import "testing"

func TestIsSubsequence(t *testing.T) {
	result := IsSubsequence("abc", "ahbgdc")
	t.Log(result)
}

func IsSubsequence(s string, t string) bool {
	if len(s) == 0 {
		return true
	}

	if len(t) == 0 {
		return true
	}

	j := 0
	for i := 0; i < len(t); i++ {
		if t[i] == s[j] {
			j++
			if j >= len(s) {
				break
			}
		}
	}

	return j == len(s)
}

File: src/python/challenges/problems/is_subsequence_test.py
# https://leetcode.com/problems/is-subsequence

from core.problem_base import *

class IsSubsequence(ProblemBase):
    def Solution(self, s: str, t: str) -> bool:
        if not s:
            return True

        if not t:
            return False

        j = 0
        for i in range(len(t)):
            if t[i] == s[j]:
                j += 1
                if j >= len(s):
                    break

        return j == len(s)

if __name__ == '__main__':
    TestGen(IsSubsequence) \
        .Add(lambda tc: tc.Param("abc").Param("ahbgdc").Result(True)) \
        .Add(lambda tc: tc.Param("axc").Param("ahbgdc").Result(False)) \
        .Run()
